File: Engine/src/main/java/akkaLabs/ExtremeAircrafts/util/PositionUtil.java
package akkaLabs.ExtremeAircrafts.util;

import akkaLabs.ExtremeAircrafts.position.Position;

/**
 * Created by Ron on 12/06/2017.
 */
public class PositionUtil {
    public static Position calculate(double distance,double heading,Position start){
        double brngRad = Math.toRadians(heading);
        double latRad = Math.toRadians(start.getLatitude());
        double lonRad = Math.toRadians(start.getLongitude());
        int earthRadiusInMetres = 6371000;
        double distFrac = distance/earthRadiusInMetres;

        double latitudeResult = Math.toDegrees(Math.asin(Math.sin(latRad) * Math.cos(distFrac) + Math.cos(latRad) * Math.sin(distFrac) * Math.cos(brngRad)));
        double a = Math.atan2(Math.sin(brngRad) * Math.sin(distFrac) * Math.cos(latRad), Math.cos(distFrac) - Math.sin(latRad) * Math.sin(latitudeResult));
        double longitudeResult =  Math.toDegrees((lonRad + a + 3 * Math.PI) % (2 * Math.PI) - Math.PI);

        return new Position(longitudeResult,latitudeResult,start.getAltitude());
    }
}

File: ScalaEngine/src/main/scala/position/PositionUtil.scala
package position

import scala.math.{toRadians,toDegrees,sin,asin,cos,atan2,Pi}
/**
  * Created by Ron on 23/06/2017.
  */
object PositionUtil {
  def calculate(distance: Double, heading: Double, start: Position) = {
    val bearingRad = toRadians(heading)
    val latRad = toRadians(start.latitude)
    val lonRad = toRadians(start.longitude)
    val earthRadiusInMetres = 6371000
    val distFrac = distance / earthRadiusInMetres
    val latitudeResult = toDegrees(asin(Math.sin(latRad) * cos(distFrac) + cos(latRad) * sin(distFrac) * cos(bearingRad)))
    val a = atan2(sin(bearingRad) * sin(distFrac) * cos(latRad), cos(distFrac) - sin(latRad) * sin(latitudeResult))
    val longitudeResult = toDegrees((lonRad + a + 3 * Pi) % (2 * Pi) - Pi)
    new Position(longitudeResult, latitudeResult, start.altitude)
  }
}
